RecoEgamma/EgammaElectronProducers/plugins/GsfElectronBaseProducer.cc

```diff
@@ -382,9 +381,7 @@ void GsfElectronBaseProducer::fillEvent( edm::Event & event )
     algo_->displayInternalElectrons("GsfElectronAlgo Info (after amb. solving)") ;
    }
   // final filling
-  auto finalCollection = std::make_unique<GsfElectronCollection>();
-  algo_->copyElectrons(*finalCollection) ;
-  orphanHandle_ = event.put(std::move(finalCollection));
+  orphanHandle_ = event.emplace(electronPutToken_, std::move(algo_->electrons()));
```

Comment:
I would add here `algo_->electrons().clear()` (or `algo_->electrons() = reco::GsfElectronCollection()`) to leave the `algo_->electrons()` in a determinate state after the move (since it will be re-used in the next event).

Reply:
Actually they are not reused. The collection is part of EventData [1] which is always recreated here every event [2].

I know that's not optimal. There are some of these pointers structures like `EventData`, `ElectronData` and I think another one which persist through the events even though they are always recreated. I suggest we should move forward here for the moment because the changes are already numerous, and some other time I can do a fresh PR with the new ideas. If we want to do all the potential improvements here is will never end :D

But yes, changing the resource acquisition model such that objects which are only meant for the use in one event don't survive the event would be on the list. And you know, I think the problem here is the split between "algo" and "producer", even though a specific algo is always written for a specific producer. If we further streamline the Algo and eventually merge it with the producer things would get a lot more clear. Any reason for these numerous splits in "algo" and "producer" in CMSSW actually?

Have a nice weekend!

[1] https://github.com/cms-sw/cmssw/blob/e172bb9c8b46b0b1560a96f72df19ab927b3d804/RecoEgamma/EgammaElectronAlgos/interface/GsfElectronAlgo.h#L314
[2] https://github.com/cms-sw/cmssw/blob/e172bb9c8b46b0b1560a96f72df19ab927b3d804/RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc#L499